data_steward/cdr_cleaner/cleaning_rules/aggregate_zip_codes.py

```diff
@@ -0,0 +1,235 @@
+"""
+To further obfuscate participant identity, some generalized zip codes will be aggregated together.
+The PII zip code and state will be transformed to a neighboring zip code/state pair for 
+those zip codes with low population density.
+It is expected that this lookup table will be static and will remain unchanged. 
+It is based on US population, and not on participant address metrics.
+
+Original Issues: DC-1379
+"""
+
+# Python imports
+import logging
+
+# Project imports
+import constants.bq_utils as bq_consts
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+from cdr_cleaner.cleaning_rules.generalize_zip_codes import GeneralizeZipCodes
+from cdr_cleaner.cleaning_rules.deid.string_fields_suppression import StringFieldsSuppression
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+from common import JINJA_ENV, OBSERVATION, PIPELINE_TABLES
+from utils import pipeline_logging
+
+LOGGER = logging.getLogger(__name__)
+PII_STATE_VOCAB = 'pii_state_vocab'
+ZIP_CODE_AGGREGATION_MAP = 'zip_code_aggregation_map'
+ZIP_CODES_AND_STATES_TO_MODIFY = '_zip_codes_and_states_to_modify'
+
+FIND_ZIP_CODES_AND_STATES_QUERY = JINJA_ENV.from_string("""
+SELECT
+    zip_code_obs.person_id, zip_code_obs.observation_id zip_code_observation_id,
+    state_obs.observation_id state_observation_id, 
+    map.zip_code_3, map.state,
+    map.transformed_zip_code_3, map.transformed_state
+FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` zip_code_obs
+JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}` map
+    ON map.zip_code_3 = zip_code_obs.value_as_string
+JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{pii_state_vocab}}` state_vocab
+    ON state_vocab.postal_code = map.state
+LEFT JOIN `{{project_id}}.{{dataset_id}}.{{obs_table}}` state_obs
+    ON state_obs.person_id = zip_code_obs.person_id
+        AND state_obs.observation_source_concept_id = 1585249
+        AND state_vocab.concept_id = state_obs.value_source_concept_id
+WHERE zip_code_obs.observation_source_concept_id = 1585250
+""")
+
```

Comment:
This is totally up to you, and may be a conversation we bring up with the other devs, but instead of creating a new table, this can be a subquery in the queries below, under the `WITH AS`

so `MODIFY_ZIP_CODES_QUERY` would look like:

```
MODIFY_ZIP_CODES_QUERY = JINJA_ENV.from_string("""
WITH modified_zip_codes_and_states AS (
SELECT
    zip_code_obs.person_id, zip_code_obs.observation_id zip_code_observation_id,
    state_obs.observation_id state_observation_id, 
    map.zip_code_3, map.state,
    map.transformed_zip_code_3, map.transformed_state
FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` zip_code_obs
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}` map
    ON map.zip_code_3 = zip_code_obs.value_as_string
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{pii_state_vocab}}` state_vocab
    ON state_vocab.postal_code = map.state
LEFT JOIN `{{project_id}}.{{dataset_id}}.{{obs_table}}` state_obs
    ON state_obs.person_id = zip_code_obs.person_id
        AND state_obs.observation_source_concept_id = 1585249
        AND state_vocab.concept_id = state_obs.value_source_concept_id
WHERE zip_code_obs.observation_source_concept_id = 1585250
)
SELECT
    obs.* REPLACE (
        COALESCE(mzc.transformed_zip_code_3, obs.value_as_string) AS value_as_string
    )
FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` obs
LEFT JOIN modified_zip_codes_and_states mzc
    ON mzc.zip_code_observation_id = obs.observation_id
        AND mzc.person_id = obs.person_id
""")
```

and `MODIFY_STATES_QUERY` would look like:

```
MODIFY_STATES_QUERY = JINJA_ENV.from_string("""
WITH modified_zip_codes_and_states AS (
SELECT
    zip_code_obs.person_id, zip_code_obs.observation_id zip_code_observation_id,
    state_obs.observation_id state_observation_id, 
    map.zip_code_3, map.state,
    map.transformed_zip_code_3, map.transformed_state
FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` zip_code_obs
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}` map
    ON map.zip_code_3 = zip_code_obs.value_as_string
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{pii_state_vocab}}` state_vocab
    ON state_vocab.postal_code = map.state
LEFT JOIN `{{project_id}}.{{dataset_id}}.{{obs_table}}` state_obs
    ON state_obs.person_id = zip_code_obs.person_id
        AND state_obs.observation_source_concept_id = 1585249
        AND state_vocab.concept_id = state_obs.value_source_concept_id
WHERE zip_code_obs.observation_source_concept_id = 1585250
)
SELECT
    obs.* REPLACE (
        COALESCE(state_vocab.concept_id, obs.value_source_concept_id) AS value_source_concept_id,
        COALESCE(state_vocab.concept_id, obs.value_as_concept_id) AS value_as_concept_id,
        COALESCE(state_vocab.concept_code, obs.value_source_value) AS value_source_value
    )
FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` obs
LEFT JOIN modified_zip_codes_and_states mzc
    ON mzc.state_observation_id = obs.observation_id
        AND mzc.person_id = obs.person_id
LEFT JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{pii_state_vocab}}` state_vocab
    ON state_vocab.postal_code = mzc.transformed_state
""")
```

Reply:
I tried refactoring the code like this, but then discovered that `MODIFY_STATES_QUERY` was operating on the transformed zip code observations by `MODIFY_ZIP_CODES_QUERY` instead of the original. I'll try doing one big CTE instead and see if that works.